python/src/scipp/coords.py

```diff
@@ -0,0 +1,151 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock, Jan-Lukas Wynen
+
+import inspect
+from typing import Union, List, Dict, Tuple, Callable
+from . import Variable, DataArray, Dataset, bins, VariableError
+
+
+def _consume_coord(obj, name):
+    if name in obj.coords:
+        obj.attrs[name] = obj.coords[name]
+        del obj.coords[name]
+    return obj.attrs[name]
+
+
+def _produce_coord(obj, name):
+    if name in obj.attrs:
+        obj.coords[name] = obj.attrs[name]
+        del obj.attrs[name]
+    return obj.coords[name]
+
+
+class CoordTransform:
+    Graph = Dict[Union[str, Tuple[str, ...]], Union[str, Callable]]
+
+    def __init__(self, obj):
+        self.obj = obj
+        self._events_copied = False
+
+    def _add_event_coord(self, key, coord):
+        try:
+            self.obj.bins.coords[key] = coord
+        except VariableError:  # Thrown on mismatching bin indices, e.g. slice
+            self.obj.data = self.obj.data.copy()
+            self.obj.bins.coords[key] = coord
+
+    def _add_event_coords(self, coords):
+        for key, coord in coords.items():
+            # Non-binned coord should be duplicate of dense handling above,
+            # if present => ignored.
+            if coord.bins is not None:
+                self._add_event_coord(key, coord)
+
+    def _add_coord(self, *, name, graph, rename):
```

Comment:
`rename` could be an instance member instead of an in/out argument. Not sure if that would be cleaner, though.